### Step 1:

In [2]:
#Import libraries
import pandas as pd
import numpy as np
import os

In [3]:
#import the orders_products_combined
ords_prods_merge = pd.read_pickle(r'C:\Users\Sthif\01-2025-Instacart Basket Analysis\02 Data\Prepared Data\orders_products_combined.pkl')
path = r'C:\Users\Sthif\01-2025-Instacart Basket Analysis'
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_combined.pkl'))

In [4]:
#Check the shape of the imported dataframe 
ords_prods_merge.shape

(32406041, 13)

In [5]:
df = ords_prods_merge[:1000000]

In [6]:
# Define a function & apply it
def price_label(row):

  if row['prices'] <= 5:
    return 'Low-range product'
  elif (row['prices'] > 5) and (row['prices'] <= 15):
    return 'Mid-range product'
  elif row['prices'] > 15:
    return 'High range'
  else: return 'Not enough data'

In [7]:
df['price_range'] = df.apply(price_label, axis=1)

C:\Users\Sthif\AppData\Local\Temp\ipykernel_6124\2848743372.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_range'] = df.apply(price_label, axis=1)


In [33]:
df = ords_prods_merge [:1000000]
df.shape

(1000000, 16)

In [16]:
# Fixing the issue
df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [10]:
df.loc[(df['prices'] <= 15) & (df['prices'] > 5), 'price_range_loc'] = 'Mid-range product' 

In [11]:
df.loc[df['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [34]:
df['price_range_loc'].value_counts(dropna = False)

KeyError: 'price_range_loc'

In [13]:
# Check for errors
df.loc[df['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [14]:
print(df['price_range_loc'].value_counts())

price_range_loc
Mid-range product     673194
Low-range product     314381
High-range product     12425
Name: count, dtype: int64


In [17]:
# Create a new column in ords_prods_merge
ords_prods_merge['orders_day_of_week'].value_counts(dropna = False)

orders_day_of_week
0    6204404
1    5660456
6    4496635
2    4213986
5    4205906
3    3840701
4    3783953
Name: count, dtype: int64

In [18]:
# Initialize result list
result = []

for value in ords_prods_merge["orders_day_of_week"]:
  if value == 0:
    result.append("Busiest day")
  elif value == 4:
    result.append("Least busy")
  else:
    result.append("Regularly busy")

In [19]:
ords_prods_merge['busiest_day'] = result
result

['Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy

In [35]:
#Frequency of the new coolumn
ords_prods_merge['busiest_day'].value_counts(dropna = False)

busiest_day
Regularly busy    22417684
Busiest day        6204404
Least busy         3783953
Name: count, dtype: int64

### Step2:

In [21]:
# Calculate the frequency of orders by day
print(ords_prods_merge.columns)

Index(['order_id', 'user_id', 'order_number', 'orders_day_of_week',
       'order_hour_of_day', 'days_since_prior_order', 'product_id',
       'add_to_cart_order', 'reordered', 'product_name', 'aisle_id',
       'department_id', 'prices', 'busiest_day'],
      dtype='object')


In [22]:
day_frequencies = ords_prods_merge['orders_day_of_week'].value_counts()
print(day_frequencies)

orders_day_of_week
0    6204404
1    5660456
6    4496635
2    4213986
5    4205906
3    3840701
4    3783953
Name: count, dtype: int64


In [23]:
# Calculate the frequency of orders by day
day_frequencies = ords_prods_merge['orders_day_of_week'].value_counts()

# Identify the two busiest and two slowest days
busiest_days = day_frequencies.nlargest(2).index.tolist()
slowest_days = day_frequencies.nsmallest(2).index.tolist()

print("Busiest days:", busiest_days)
print("Slowest days:", slowest_days)

Busiest days: [0, 1]
Slowest days: [4, 3]


In [24]:
# Define a function
def classify_days(row, busiest_days, slowest_days):
    if row['orders_day_of_week'] in busiest_days:
        return "Busiest days"
    elif row['orders_day_of_week'] in slowest_days:
        return "Slowest days"
    else:
        return "Regularly busy"

In [25]:
ords_prods_merge['day_summary'] = ords_prods_merge.apply(lambda row: classify_days(row, busiest_days, slowest_days), axis=1)

### Step 3:

In [26]:
# Check the values in the new column
day_summary_frequency = ords_prods_merge['day_summary'].value_counts()
print("Frequency of day_summary column:")
print(day_summary_frequency)

Frequency of day_summary column:
day_summary
Regularly busy    12916527
Busiest days      11864860
Slowest days       7624654
Name: count, dtype: int64


###### Busiest days account for 11,864,860 rows, highlighting the high activity on these days.
###### Slowest days account for 7,624,654 rows, indicating significantly lower activity.
###### Regularly busy rows (12,916,527) dominate the dataset, suggesting consistent activity across these days.

### Step 4:

In [27]:
# Calculate the frequency of orders by hour
hour_frequencies = ords_prods_merge['order_hour_of_day'].value_counts()

In [28]:
# Determine thresholds for "Most orders," "Average orders," and "Fewest orders"
most_orders = hour_frequencies.nlargest(8).index.tolist()  # Top 8 hours
fewest_orders = hour_frequencies.nsmallest(8).index.tolist()  # Bottom 8 hours

In [29]:
# Define a function to classify the time periods
def classify_period(hour):
    if hour in most_orders:
        return "Most orders"
    elif hour in fewest_orders:
        return "Fewest orders"
    else:
        return "Average orders"

In [30]:
# Apply the function to create the new column
ords_prods_merge['busiest_period_of_day'] = ords_prods_merge['order_hour_of_day'].apply(classify_period)

### Step 5:

In [31]:
# Print the frequency of the 'busiest_period_of_day' column
busiest_period_frequency = ords_prods_merge['busiest_period_of_day'].value_counts()
print("Frequency of busiest_period_of_day column:")
print(busiest_period_frequency)

Frequency of busiest_period_of_day column:
busiest_period_of_day
Most orders       21118857
Average orders     9997990
Fewest orders      1289194
Name: count, dtype: int64


### Step 7:

In [32]:
# Check for missing values
ords_prods_merge.isnull().sum()

order_id                        0
user_id                         0
order_number                    0
orders_day_of_week              0
order_hour_of_day               0
days_since_prior_order    2076201
product_id                      0
add_to_cart_order               0
reordered                       0
product_name                    0
aisle_id                        0
department_id                   0
prices                          0
busiest_day                     0
day_summary                     0
busiest_period_of_day           0
dtype: int64

In [40]:
# Drop rows where 'days_since_prior_order' has missing values
ords_prods_merge.dropna(subset=['days_since_prior_order'], inplace=True)

In [41]:
# Check if all missing values are gone
print(ords_prods_merge['days_since_prior_order'].isnull().sum()) 

0


In [44]:
# Check for duplicates in the merged dataset
ords_prods_merge.duplicated().sum()
duplicates = ords_prods_merge[ords_prods_merge.duplicated()]
print(duplicates)

          order_id  user_id  order_number  orders_day_of_week  \
35812      2564286      236            32                   1   
35832      1759951      236            33                   4   
154303     2508582     1021            30                   2   
164231      265038     1095            27                   3   
164264     2226059     1095            28                   5   
...            ...      ...           ...                 ...   
32373376   2897910   205987             7                   3   
32383922    927735   206059             2                   3   
32383949   2627635   206059             4                   4   
32384086    163963   206059            16                   2   
32395391   2965594   206126            25                   2   

          order_hour_of_day  days_since_prior_order  product_id  \
35812                    23                     2.0       35306   
35832                     0                     3.0       35306   
154303            

In [45]:
ords_prods_merge.drop_duplicates(inplace=True)

In [49]:
# Check the number of duplicates remaining
print(ords_prods_merge.duplicated().sum())

0


In [62]:
# Convert to category
ords_prods_merge['busiest_day'] = ords_prods_merge['busiest_day'].astype('category')
ords_prods_merge['day_summary'] = ords_prods_merge['day_summary'].astype('category')
ords_prods_merge['busiest_period_of_day'] = ords_prods_merge['busiest_period_of_day'].astype('category')

In [64]:
# Export Data in Pickle Format
ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_combined.pkl'))